File: base/reverse/reverse.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import unittest

def reverse(str):
    stack = []
    for c in str:
        stack += [c]
    reversed_str = ""
    while True:
        if stack:
            reversed_str += stack.pop()
            continue
        return reversed_str

def revrec(input, output):
    if input == "":
        return output
    else:
        c = input[0]
        rest = input[1:]
        return revrec(rest, c+output)

print revrec("hoge", "")
print revrec("hoge", "")

class TestMerge(unittest.TestCase):
    def test_solve(self):
        input    = "foobar"
        expected = "raboof" 
        self.assertEqual(reverse(input), expected)
if __name__ == '__main__':
    unittest.main()

File: base/reverse/reverse.rb
def rev(str)
    stack = []
    output = ""
    str.each_char do |c|
        stack.unshift(c)
    end
    for c in stack do 
        output << c
    end
    output
end

def revrec(str)
    def rev_l(head, rest) 
        if rest == ""
            head
        else
            rev_l(rest[0]+head, rest[1..-1])
        end
    end
    if str.length <= 1
        str
    else
        rev_l(str[0], str[1..-1])
    end
end

p revrec("piyopiyo")
p rev("piyopiyo")
